06-trees/tree-answers.txt

```diff
@@ -0,0 +1,22 @@
+Based on my implementation, it's hard to tell when you would use
+a binary heap.
+
+Based on other resources, the time insertion for a binary
+heap should be faster on average. Binary heaps can also be efficiently
+implemented on top of arrays, so you wouldn't need 3 pointers per node
+like a binary search tree requires. The binary search tree is better for
+general searching; searching in a binary heap is only efficient for
+searching the max/min.
```

Comment:
The main thing I see in your code is an excess of `nil` checks and some refactoring that could be done to reduce the number of conditionals in general. For moving a node up, the only `nil` check that is required is to see if there is a parent at all. For moving down, you may check for the `left` to be `nil` first; if it is, then you do not even need to bother checking the `right`. If it is not, checking the `right` will yield whether or not the check for which one is smaller is even necessary. The only check you need for `node.parent.nil?` is to ensure the `@root` is changed appropriately. So, the code could look something like this (disclaimer - I am replicating my solution from memory, as I do not have mine handy at the moment):

```ruby
def move_down(node)
  unless node.left.nil?
    if node.right.nil? || node.left.rating < node_right.rating
      # Do the typical node swap
      # * node.left.right = node.right is fine even if `node.right` is `nil`
      # * node.left.parent = node.parent is fine even if `node.parent` is `nil`
      # Change the typical line changing right's parent to this: node_right.parent = node_left unless node_right.nil?
      # Add the check for if `node.left` became root
    else
      # Do the typical swap for node and node.right
      # Add the check for if `node.right` became root
    end
  end
end
```

While not an optimization for insertion, it does showcase how more concise the code can be without so many `nil` checks. All of those extra conditionals slow down the algorithm for insertion tremendously.